File: Codes/Programmers/수식최대화.py
def calculate(order, arr):
    for o in order:
        nxt = []
        i = 0
        while i < len(arr):
            now = arr[i]
            if now == o:
                nxt.append(str(eval(nxt.pop() + now + arr[i+1])))
                i += 2
            else:
                nxt.append(now)
                i += 1
        arr = nxt
    return abs(int(arr.pop()))
        

def solution(expr):
    result = 0
    splited = ['']
    for e in expr:
        if e.isdigit():
            splited[-1] += e
        else:
            splited.append(e)
            splited.append('')
    
    all_orders = ['-*+', '-+*', '*-+', '+-*', '*+-', '+*-']
    
    for order in all_orders:
        result = max(calculate(order, splited), result)
        
    return result

File: Codes/Programmers/수식최대화.js
function parseExpression(expression) {
  let nowNumber = "";
  const formula = [];
  for (let char of expression) {
    const charCode = char.charCodeAt();
    if ("0".charCodeAt() <= charCode && charCode <= "9".charCodeAt()) {
      nowNumber += char;
    } else {
      formula.push(parseInt(nowNumber));
      formula.push(char);
      nowNumber = "";
    }
  }
  formula.push(parseInt(nowNumber));
  return formula;
}

function operate(num1, num2, operator) {
  if (operator === "+") {
    return num1 + num2;
  } else if (operator === "-") {
    return num1 - num2;
  } else {
    return num1 * num2;
  }
}

function calculate(formula, priority) {
  let now = [...formula];
  for (let operator of priority) {
    let i = 0;
    let temp = [];
    while (i < now.length) {
      if (now[i] === operator) {
        temp.push(operate(temp.pop(), now[i + 1], operator));
        i++;
      } else {
        temp.push(now[i]);
      }
      i++;
    }
    now = temp;
  }
  return Math.abs(now[0]);
}

function solution(expression) {
  let answer = 0;
  const formula = parseExpression(expression);
  const priorityList = ["+*-", "+-*", "-*+", "-+*", "*-+", "*+-"];
  priorityList.forEach((priority) => {
    answer = Math.max(calculate(formula, priority), answer);
  });

  return answer;
}
